File: src/Tetris.java
import java.util.Timer;
import java.util.TimerTask;

class Tetris {
    private Field field = new Field();
    private boolean pause = false;
    private boolean stopped = true;
    private int level = 1000;
    private int period = 1000;
    private int delay = 1000;
    private int score = 0;
    private int difficulty = 0;
    private double acceleration = 0.98;
    private boolean gameOver = false;
    private Timer timer;

    void setDifficulty(int difficulty) {
        this.difficulty = difficulty;
    }

    void start() {
        field.spawnShape();
        acceleration = 1 - (difficulty + 1) * 0.02;
        level = 1000 - difficulty * 100;
        period = level;
        delay = level;
        stopped = false;
        Controller.refresh();
        startTimer();
    }

    boolean isStopped() {
        return stopped;
    }

    void figureFall() {
        if (stopped || pause || gameOver || period <= 50) return;
        period = 50;
        delay = 0;
        stopTimer();
        startTimer();
    }

    boolean isGameOver() {
        return gameOver;
    }

    Integer getScore() {
        return score;
    }

    private synchronized void moveDown() {
        if (pause || gameOver || stopped) return;
        if (!field.moveDown()) {
            int removedLines = field.removeFullLines();
            score += removedLines;
            gameOver = !field.spawnShape();
            if (removedLines > score % 5) {
                for (int i = 0; i < removedLines / 5 + ((score - removedLines) % 5 + removedLines % 5) / 5; i++) {
                    level = (int) (level * acceleration);
                }
                period = level;
                delay = level;
                stopTimer();
                startTimer();
            }
        }
        Controller.refresh();
    }

    int getCellValue(int y, int x) {
        return field.getCellValue(y, x);
    }

    int getNextShapeCellValue(int y, int x) {
        return field.getNextShapeCellValue(y, x);
    }

    int getNextShapeNumber() {
        return field.getNextShapeNumber();
    }

    boolean isPause() {
        return pause;
    }

    synchronized void moveLeft() {
        if (pause || gameOver || stopped) return;
        field.moveLeft();
        Controller.refresh();
    }

    synchronized void moveRight() {
        if (pause || gameOver || stopped) return;
        field.moveRight();
        Controller.refresh();
    }

    synchronized void rotate() {
        if (pause || gameOver || stopped) return;
        field.rotateShape();
        Controller.refresh();
    }

    void stop() {
        field.clear();
        stopTimer();
        score = 0;
        stopped = true;
        gameOver = false;
        pause = false;
        Controller.refresh();
    }

    void restart() {
        stop();
        start();
    }

    void figureRestore() {
        if (gameOver || stopped || pause || period == level) return;
        period = level;
        delay = level;
        stopTimer();
        startTimer();
    }

    void pause() {
        if (gameOver || stopped) return;
        pause = !pause;
        if (pause) {
            stopTimer();
        } else {
            startTimer();
        }
        Controller.refresh();
    }

    private void stopTimer() {
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
    }

    private void startTimer() {
        if (timer != null) {
            timer = null;
        }
        timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                moveDown();
                if (gameOver) {
                    timer.cancel();
                    timer = null;
                    Controller.refresh();
                }
            }
        }, delay, period);
    }
}

File: src/game/Tetris.kt
package game

import java.util.*

class Tetris {
    private val field = Field()
    private var level = 1000
    private var period = 1000
    private var delay = 1000
    private var difficulty = 0
    private var acceleration = 0.98
    private var timer: Timer? = null
    val nextShapeNumber: Int
        get() = this.field.nextShapeNumber
    var isPause = false
        private set
    var isStopped = true
        private set
    var score = 0
        private set
    var isGameOver = false
        private set
    var readyToSolve: Boolean
        set(value) {
            this.field.shapeSolved = !value
        }
        get() = !this.field.shapeSolved
    val statesCount: Int
        get() = this.field.getStatesCount()

    fun getShapeNeighbors(st: Int): Array<IntArray> {
        return field.getShapeNeighbors(st)
    }

    fun getBottomCells(st: Int) = field.getBottomCells(st)

    fun horizontalMoveRange(st: Int): Int {
        return field.getHorizontalMoveRange(st)
    }


    fun getProjection(offX: Int, st: Int) = field.projection(offX, st)

    fun getFallHeight(offX: Int, st: Int): Int {
        return field.height(offX, st)
    }

    fun setDifficulty(difficulty: Int) {
        this.difficulty = difficulty
    }

    fun start() {
        field.spawnShape()
        acceleration = 1 - (difficulty + 1) * 0.02
//        level = 1000 - difficulty * 99
        level = 1
        period = level
        delay = level
        isStopped = false
        startTimer()
    }

    fun figureFall() {
        if (isStopped || isPause || isGameOver || period <= 50) return
        period = 50
        delay = 0
        stopTimer()
        startTimer()
    }

    @Synchronized
    private fun moveDown() {
        if (isPause || isGameOver || isStopped) return
        if (!field.moveDown()) {
            val removedLines = field.removeFullLines()
            score += removedLines
            isGameOver = !field.spawnShape()
//            if (removedLines > score % 5) {
//                for (i in 0 until removedLines / 5 + ((score - removedLines) % 5 + removedLines % 5) / 5) {
//                    level = (level * acceleration).roundToInt()
//                }
//                period = level
//                delay = level
//                stopTimer()
//                startTimer()
//            }
        }
    }

    fun getCellValue(y: Int, x: Int): Int {
        return field.getCellValue(y, x)
    }

    fun getNextShapeCellValue(y: Int, x: Int): Int {
        return field.getNextShapeCellValue(y, x)
    }

    @Synchronized
    fun moveLeft(): Boolean {
        if (isPause || isGameOver || isStopped) return false
        return field.moveLeft()
    }

    @Synchronized
    fun moveRight(): Boolean {
        if (isPause || isGameOver || isStopped) return false
        return field.moveRight()
    }

    @Synchronized
    fun rotate(): Boolean {
        if (isPause || isGameOver || isStopped) return false
        return field.rotateShape()
    }

    fun stop() {
        field.clear()
        stopTimer()
        score = 0
        isStopped = true
        isGameOver = false
        isPause = false
    }

    fun restart() {
        stop()
        start()
    }

    fun figureRestore() {
        if (isGameOver || isStopped || isPause || period == level) return
        period = level
        delay = level
        stopTimer()
        startTimer()
    }

    fun pause() {
        if (isGameOver || isStopped) return
        isPause = !isPause
        if (isPause) {
            stopTimer()
        } else {
            startTimer()
        }
    }

    private fun stopTimer() {
        if (timer != null) {
            timer!!.cancel()
            timer = null
        }
    }

    private fun startTimer() {
        if (timer != null) {
            timer = null
        }
        timer = Timer()
        timer!!.schedule(object : TimerTask() {
            override fun run() {
                moveDown()
                if (isGameOver) {
                    timer!!.cancel()
                    timer = null
                }
            }
        }, delay.toLong(), period.toLong())
    }
}
